### PR TITLE
Build documentation using nix for CI

### PR DESCRIPTION
**Pull request by afh**
This PR introduces `flake.nix`, which is used to build the [website](http://github.com/ledger/ledger.github.io) and documentation for [ledger](http://github.com/ledger/ledger) and [ledger-mode](http://github.com/ledger/ledger-mode).
The flake contains 3 derivations:
1. `website` (the default) to build the static content for ledger-cli.org
2. `doc-ledger` to build the HTML and PDF manual and man-page for ledger
3. `doc-ledger-mode` to build the HTML and PDF manual and man-page for ledger-mode

and is used by the build-and-deploy workflow to build the website and documentation and deploy it to ledger-cli.org.

**NOTA BENE:** The flake references ledger and ledger-mode releases to build the documentation, hence when a new version of ledger or ledger-mode is released the tag needs to be updated in the flake (see [`inputs` in flake.nix](https://github.com/ledger/ledger.github.io/pull/63/files#diff-206b9ce276ab5971a2489d75eb1b12999d4bf3843b7988cbe8d687cfde61dea0R15-R17)) and `nix flake update` run.

### CODE REVIEW

**Comment by afh** [comment]
The changes in this PR are now ready for review, but should not me merged until #46 is resolved.

**Comment by afh** [comment]
Those that are not using nix and would like to test the result of the build consider downloading a [github-pages build artifact](https://github.com/ledger/ledger.github.io/suites/11701786216/artifacts/608737300) from one of the [GitHub Action runs](https://github.com/ledger/ledger.github.io/actions/runs/4478279933), unpack it, start a local webserver within the unpacked directory: `% python3 -m http.server || python -m SimpleHTTTPServer`, and visit http://localhost:8000

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+          #version = "3.3.1";
+          #hash = "sha256-gBqf0ty8je912pZoupcA099515ngIK8BK0uk8nXz9M0=";
+          # TODO: Remove once ledger release contains changes from afh/ledger@web-docs
+          version = "7b7bb08f5c5903566f866af23acde6a3c451b7d3";
+          hash = "sha256-lZdHdjFO/RhtKX8/4QrnLpeF+cbXVkzz9JvNF6DAWE0=";
+        };
+        ledger-mode = {
+          version = "4.0.0";
+          hash = "sha256-YF7JfTe2GVWgb14XghZG0djbMGVzYtZmyjcBvHL02do=";
+        };
```
Comment (purcell) [inline flake.nix:28]: I expect these should just be flake inputs. You can use tags there, for example.

**Review by purcell — APPROVED** [review]
Approving because it looks generally sane to me, but I've suggested some changes.

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+         cmakeFlags = [ "-DBUILD_WEB_DOCS:BOOL=ON" "-Wno-dev" ];
+
+         buildFlags = "doc";
+       };
+
+       doc-ledger-mode = pkgs.stdenvNoCC.mkDerivation rec {
+         name = "ledger-mode";
+         version = dist."${name}".version;
+
+         src = pkgs.fetchFromGitHub {
```
[the remark this answers is not in the document]
Comment (afh) [reply flake.nix:97]: I'm not sure how to get the source from an `input`. With the changes to the inputs (_v.s._) I've tried `src = import ledger;` and `src = import ledger { inherit ledger; };` yet that fails with `error: getting status of '/nix/store/jp479mqs3j46s0n63n5k0i4ck4qijqh3-source/flake.nix': No such file or directory`.

Can you point me to an example, @purcell?

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+         src = ledger;
+
+         nativeBuildInputs = with pkgs; [
+           cmake groff texinfo tex
+         ];
+
+         enableParallelBuilding = false;
+
+         preConfigure = "cd doc";
```
Comment (afh) [inline flake.nix:68]: @purcell using `sourceRoot = "${src}/doc";` fails with  `chmod: changing permissions of '/nix/store/3i2ml3yk9h9dqqhg5hlixawd86qjyvvs-source/doc': Operation not permitted` for several files in the source. Using `preConfigure = "cd doc";` works, but I was wondering if there is a "nicer" solution (possibly involving `sourceRoot` or something similar). Would you happen to have an idea?

Comment (afh) [reply flake.nix:68]: Thanks for your input, @purcell, much appreciated. Happy to keep it as is until a "nicer" solution is found.

**Comment by purcell** [comment]
Off-hand observation: overall it's slightly weird for a flake in a repo to refer to a fixed revision in that same repo: you'd normally use it to build whatever is in the repo alongside the flake definition. An alternative here — which might only become feasible once there's a release which contains the flake definition — would be always build the docs from the current source, but only deploy them on a certain branch, e.g. "release" or "latest".

**Comment by purcell** [comment]
Oh never mind, I'm talking nonsense, since this flake is specifically for the doc site repo. Please disregard!

**Comment by afh** [comment]
Thanks for your thoughts, @purcell.

FWIW: I explored the possibility of having a docs package in the ledger flake and one in the ledger-mode flake, so that `nix build github:ledger/ledger/v3.3.1#doc` could be used in this repo's GitHub Action to build the ledger and ledger-mode documentation.
But it seems much easier to tie together all that is necessary to build the docs in this repo's flake rather than coordinating it across several repos. Thanks again for suggesting `inputs`, very helpful and much cleaner this way 🙂👍

**Comment by purcell** [comment]
Nice work! 🙌

**Pull request merged**